pdns/pdns_recursor.cc

```diff
@@ -653,7 +653,7 @@ LWResult::Result asendto(const char *data, size_t len, int flags,
   pair<MT_t::waiters_t::iterator, MT_t::waiters_t::iterator> chain=MT->d_waiters.equal_range(pident, PacketIDBirthdayCompare());
 
   for(; chain.first != chain.second; chain.first++) {
-    if(chain.first->key.fd > -1) { // don't chain onto existing chained waiter!
+    if(chain.first->key.fd > -1 && !chain.first->key.closed) { // don't chain onto existing chained waiter or a chain already processed
```

Comment:
So if I understand that right, we yield in `doResends()` by calling `MTasker::sendEvent()`, and a different mthread then tries to add the same query again. This change makes it so that it does not try to reuse the old one, which is good, but I guess that mthread will then call `MTasker::waitEvent()` via `arecvfrom()`. How does that not fail (`d_waiters.count(key)` != 0) if the old entry is still there? And if it does not fail, is there any risk that our iterator gets invalidated?

Reply:
I'm not sure we yield. What I've seen is add to chain, process, add to chainin the same mthread. I suspect the specific record is needed twice in the same resolution process and bcause of the TTL 0 the cached one is not used.

Reply:
I'm not sure I understand how that would happen with only one mthread. Wouldn't the entry be removed in `MTasker::sendEvent()`, called from `handleUDPServerResponse()`, before the correct mthread is woken up from the kernel one?

Reply:
If I run with the diff below on top of the PR (it disabled the !close condition and adds logging). I see 👍 
```
...
36 asendto() add to chain
36 doResends called!
35 asendto() add to chain
35 doResends called!
36 asendto() add to chain
36 doResends called!
38 asendto() add to chain
38 doResends called!
38 asendto() add to chain
...
```
So there does seem to be yielding going on...
This is with two clients running dig in a loop:
```
while true; do dig +time=5 +tries=1 -p 5301 euaz.tr.teams.microsoft.com ; sleep .05; done | grep "Query time"
```
```
diff --git a/pdns/pdns_recursor.cc b/pdns/pdns_recursor.cc
index 49cf2c463..d95be13d7 100644
--- a/pdns/pdns_recursor.cc
+++ b/pdns/pdns_recursor.cc
@@ -653,7 +653,8 @@ LWResult::Result asendto(const char *data, size_t len, int flags,
   pair<MT_t::waiters_t::iterator, MT_t::waiters_t::iterator> chain=MT->d_waiters.equal_range(pident, PacketIDBirthdayCompare());
 
   for(; chain.first != chain.second; chain.first++) {
-    if(chain.first->key.fd > -1 && !chain.first->key.closed) { // don't chain onto existing chained waiter or a chain already processed
+    if(chain.first->key.fd > -1) { // don't chain onto existing chained waiter or a chain already processed
+      cerr << MT->getTid() << " asendto() add to chain" << endl;
       /*
       cerr<<"Orig: "<<pident.domain<<", "<<pident.remote.toString()<<", id="<<id<<endl;
       cerr<<"Had hit: "<< chain.first->key.domain<<", "<<chain.first->key.remote.toString()<<", id="<<chain.first->key.id
@@ -3819,7 +3820,7 @@ static void doResends(MT_t::waiters_t::iterator& iter, PacketID resend, const st
 
   if(iter->key.chain.empty())
     return;
-  //  cerr<<"doResends called!\n";
+  cerr<< MT->getTid() << " doResends called!\n";
   for(PacketID::chain_t::iterator i=iter->key.chain.begin(); i != iter->key.chain.end() ; ++i) {
     resend.fd=-1;
     resend.id=*i;
```